src/ucp/core/ucp_worker.c

```diff
@@ -2243,69 +2257,78 @@ static unsigned ucp_worker_discard_uct_ep_destroy_progress(void *arg)
     uct_ep_h uct_ep     = req->send.discard_uct_ep.uct_ep;
     ucp_ep_h ucp_ep     = req->send.ep;
     ucp_worker_h worker = ucp_ep->worker;
-    khiter_t iter;
 
     ucp_trace_req(req, "destroy uct_ep=%p", uct_ep);
+    uct_ep_destroy(uct_ep);
 
     UCS_ASYNC_BLOCK(&worker->async);
-    ucp_worker_put_flush_req(req);
-    iter = kh_get(ucp_worker_discard_uct_ep_hash,
-                  &worker->discard_uct_ep_hash, uct_ep);
-    if (iter == kh_end(&worker->discard_uct_ep_hash)) {
-        ucs_fatal("no %p UCT EP in the %p worker hash of discarded UCT EPs",
-                  uct_ep, worker);
-    }
-    kh_del(ucp_worker_discard_uct_ep_hash, &worker->discard_uct_ep_hash, iter);
+    ucp_worker_discard_uct_ep_complete(req);
     UCS_ASYNC_UNBLOCK(&worker->async);
 
-    uct_ep_destroy(uct_ep);
-
     return 1;
 }
 
-static void
-ucp_worker_discard_uct_ep_flush_comp(uct_completion_t *self)
+static UCS_F_MAYBE_UNUSED int
+ucp_worker_discard_uct_ep_check_filter(const ucs_callbackq_elem_t *elem,
+                                       void *arg)
+{
+    /* No other callbacks should be scheduled for the same UCP request
+     * simultaneously */
+    ucs_assertv_always(elem->arg != arg,
+                       "req %p: there is already scheduled operation (%s)",
+                       elem->arg, ucs_debug_get_symbol_name(elem->cb));
+    return 0;
+}
+
+static void ucp_worker_discard_uct_ep_check_no_progress(ucp_request_t *req)
+{
+#if UCS_ENABLE_ASSERT
+    ucp_worker_h worker = req->send.ep->worker;
+
+    ucs_callbackq_remove_if(&worker->uct->progress_q,
+                            ucp_worker_discard_uct_ep_check_filter, req);
```

Comment:
seems different logic in debug mode?

Reply:
unfortunately, we don't have just checking existence, so I re-used `remove_if` function - but it will fail with `ucs_assert_always` if it finds something.